app/src/main/java/org/thoughtcrime/securesms/groups/GroupsV1MigrationUtil.java

```diff
@@ -0,0 +1,244 @@
+          Log.w(TAG, "Failed to migrate job due to the membership not yet being suitable for GV2. Aborting.", e);
+          return;
+        } catch (GroupAlreadyExistsException e) {
+          Log.w(TAG, "Someone else created the group while we were trying to do the same! It exists now. Continuing on.", e);
+        }
+        break;
+      case NOT_A_MEMBER:
+        Log.w(TAG, "The migrated group already exists, but we are not a member. Doing a local leave.");
+        handleLeftBehind(context, gv1Id, groupRecipient, threadId);
+        return;
+      case FULL_OR_PENDING_MEMBER:
+        Log.w(TAG, "The migrated group already exists, and we're in it. Continuing on.");
+        break;
+      default: throw new AssertionError();
+    }
+
+    Log.i(TAG, "Migrating local group " + gv1Id + " to " + gv2Id);
+
+    DecryptedGroup decryptedGroup = performLocalMigration(context, gv1Id, threadId, groupRecipient);
+
+    if (newlyCreated && decryptedGroup != null && !SignalStore.internalValues().disableGv1AutoMigrateNotification()) {
+      Log.i(TAG, "Sending no-op update to notify others.");
+      GroupManager.sendNoopUpdate(context, gv2MasterKey, decryptedGroup);
+    }
+  }
+
+  public static void performLocalMigration(@NonNull Context context, @NonNull GroupId.V1 gv1Id) throws IOException
+  {
+    Log.i(TAG, "Beginning local migration! V1 ID: " + gv1Id, new Throwable());
+    try (Closeable ignored = GroupsV2ProcessingLock.acquireGroupProcessingLock()) {
+      if (DatabaseFactory.getGroupDatabase(context).groupExists(gv1Id.deriveV2MigrationGroupId())) {
+        Log.w(TAG, "Group was already migrated! Could have been waiting for the lock.", new Throwable());
+        return;
+      }
+
+      Recipient recipient = Recipient.externalGroupExact(context, gv1Id);
+      long      threadId  = DatabaseFactory.getThreadDatabase(context).getOrCreateThreadIdFor(recipient);
+
+      performLocalMigration(context, gv1Id, threadId, recipient);
+      Log.i(TAG, "Migration complete! (" + gv1Id + ", " + threadId + ", " + recipient.getId() + ")", new Throwable());
+    } catch (GroupChangeBusyException e) {
+      throw new IOException(e);
+    }
+  }
+
+  private static @Nullable DecryptedGroup performLocalMigration(@NonNull Context context,
+                                                                @NonNull GroupId.V1 gv1Id,
+                                                                long threadId,
+                                                                @NonNull Recipient groupRecipient)
+      throws IOException, GroupChangeBusyException
+  {
+    Log.i(TAG, "performLocalMigration(" + gv1Id + ", " + threadId + ", " + groupRecipient.getId());
+
+    try (Closeable ignored = GroupsV2ProcessingLock.acquireGroupProcessingLock()){
+      DecryptedGroup decryptedGroup;
+      try {
+        decryptedGroup = GroupManager.addedGroupVersion(context, gv1Id.deriveV2MigrationMasterKey());
+      } catch (GroupDoesNotExistException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)